File: src/py_custom/py_custom/action_server.py
import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node

from custom_interfaces.action import Fibonacci


class FibonacciActionServer(Node):
    # 初始化FibonacciActionServer节点
    def __init__(self):
        super().__init__("fibonacci_action_server")
        self.__action_server = ActionServer(
            self, Fibonacci, "fibonacci", self.__execute
        )

    async def __execute(self, goal_handle):
        # 执行Fibonacci序列生成的逻辑
        self.get_logger().info("Executing goal...")
        sequence = [0, 1]

        if goal_handle.request.order < 0:
            # 如果请求的Fibonacci序列的阶数小于0，记录错误并中止目标
            self.get_logger().error("Invalid order, aborting goal...")
            goal_handle.abort()
            result = Fibonacci.Result()
            result.sequence = sequence
            return result

        for i in range(1, goal_handle.request.order):
            # 检查目标是否被取消，如果是则记录并返回当前结果
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Goal canceled.")
                result = Fibonacci.Result()
                result.sequence = sequence
                return result

            # 生成Fibonacci序列
            sequence.append(sequence[i] + sequence[i - 1])

            # 发布反馈信息
            feedback = Fibonacci.Feedback()
            feedback.partial_sequence = sequence
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f"Sending feedback: {sequence}")
            rclpy.spin_once(self, timeout_sec=0.1)

        # 目标成功完成
        goal_handle.succeed()

        # 设置并返回结果
        result = Fibonacci.Result()
        result.sequence = sequence
        self.get_logger().info(f"Result: {sequence}")
        return result


def main(args=None):
    # 主函数，初始化ROS2节点并运行ActionServer
    rclpy.init(args=args)
    rclpy.spin(FibonacciActionServer())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/cpp_custom/src/action_server.cpp
#include "custom_interfaces/action/fibonacci.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using custom_interfaces::action::Fibonacci;
using std::placeholders::_1;
using std::placeholders::_2;
using namespace std::chrono_literals;

// 斐波那契数列动作服务器类，继承自rclcpp::Node
class FibonacciActionServer : public rclcpp::Node {
  public:
    // 构造函数，初始化节点并创建动作服务器
    FibonacciActionServer() : Node("fibonacci_action_server") {
        this->action_server_ =
            rclcpp_action::create_server<Fibonacci>(this,
                                                    "fibonacci",
                                                    std::bind(&FibonacciActionServer::handle_goal, this, _1, _2),
                                                    std::bind(&FibonacciActionServer::handle_cancel, this, _1),
                                                    std::bind(&FibonacciActionServer::handle_accepted, this, _1));
    }

  private:
    // 动作服务器的共享指针
    rclcpp_action::Server<Fibonacci>::SharedPtr action_server_;

    // 处理目标请求的方法
    rclcpp_action::GoalResponse handle_goal(const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const Fibonacci::Goal> goal) {
        RCLCPP_INFO(this->get_logger(), "Received goal request: %d", goal->order);

        if (goal->order < 0) {
            RCLCPP_ERROR(this->get_logger(), "%d is an invalid order, order must be non-negative! Rejecting...", goal->order);
            return rclcpp_action::GoalResponse::REJECT;
        }

        RCLCPP_INFO(this->get_logger(), "Received a goal: %d", goal->order);
        RCLCPP_DEBUG(this->get_logger(), "Goal ID: %s", rclcpp_action::to_string(uuid).c_str());
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    // 处理取消请求的方法
    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>>) {
        RCLCPP_INFO(this->get_logger(), "Received cancel request.");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    // 处理已接受目标的方法
    void handle_accepted(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> goal_handle) {
        std::thread{std::bind(&FibonacciActionServer::execute, this, _1), goal_handle}.detach();
    }

    // 执行目标的动作
    void execute(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Fibonacci>> goal_handle) {
        RCLCPP_INFO(this->get_logger(), "Excuting goal...");

        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<Fibonacci::Feedback>();
        auto &sequence = feedback->partial_sequence;
        sequence.push_back(0);
        sequence.push_back(1);

        auto result = std::make_shared<Fibonacci::Result>();

        for (int i = 1; i < goal->order && rclcpp::ok(); i++) {
            if (goal_handle->is_canceling()) {
                result->sequence = sequence;
                goal_handle->canceled(result);
                RCLCPP_INFO(this->get_logger(), "Goal canceled.");
                return;
            }

            sequence.push_back(sequence[i] + sequence[i - 1]);
            goal_handle->publish_feedback(feedback);
            RCLCPP_INFO(this->get_logger(), "Publishing feedback...");
            std::this_thread::sleep_for(200ms);
        }

        if (rclcpp::ok()) {
            result->sequence = sequence;
            goal_handle->succeed(result);
            RCLCPP_INFO(this->get_logger(), "Goal succeeded.");
        }
    }
};

// 主函数，初始化ROS 2，创建并运行动作服务器，最后关闭ROS 2
int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<FibonacciActionServer>());
    rclcpp::shutdown();
    return 0;
}